mathics/builtin/arithmetic.py

```diff
@@ -79,7 +79,10 @@
 from mathics.eval.nevaluator import eval_N
 
 
-@lru_cache(maxsize=4096)
+# Caching this function produce problems tracking
+# precision.
+#
+# @lru_cache(maxsize=4096)
```

Comment:
Interesting - caching considered bad. Well, actually it is _improper_ caching.   A fix here is to see if one can add in the current precision to the cache.  This would be done inside the code rather than as a decorator on the code. However the lru_cache package would be used (rather than some homegrown code like we have for Expression).

Reply:
Yes, it could be implemented in that way. This is why I added a comment instead of erasing the line. By now, I think it is enough to fix the issue and improve the performance. Then, we can write another patch to restore the cache having into account the precision. If you want, I can do that in this same branch.

Reply:
Ok  that makes sense. 

But I would ask that we add a unit test for the bug here. 

That in general is good practice for any bugs that come up and get fixed. Here, it would be a test that precision is not altered across any Mathics operation that changes the mpmath precisiono.

Reply:
I added a bunch of basic tests now. If you have more in mind, please add over them.

Reply:
Also, I could reinstall the lru_cache. Doing that, I think I got a cleaner code, which (even without the cache) is even faster.